server: job status endpoint crashes node

I noticed some funky jobs on the UI and tried to test out the debug endpoint to get more details but attempting to load it crashed the node. Looking at the logs:

```
panic: attempt to clone *jobspb.Payload, which contains uncloneable field of type jobspb.isPayload_Details

goroutine 870200 [running]:
panic(0x3c74880, 0xc09f0c01d0)
        /usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc000c79590 sp=0xc000c79500 pc=0x79a3d2
github.com/cockroachdb/cockroach/pkg/util/protoutil.Clone(0x4ba0960, 0xc00eb78800, 0xc000c795f0, 0xc000c795e8)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/protoutil/clone.go:77 +0x173 fp=0xc000c795f8 sp=0xc000c79590 pc=0xf91563
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).ToProto(0xc00eb787e0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:575 +0xc3 fp=0xc000c79670 sp=0xc000c795f8 pc=0x16fb753
github.com/cockroachdb/cockroach/pkg/server.(*statusServer).JobStatus(0xc00028d6c0, 0x4b67ae0, 0xc05cc554a0, 0xc0928ac0e0, 0xc00028d6c0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/status.go:2072 +0xf7 fp=0xc000c796c8 sp=0xc000c79670 pc=0x30580d7
github.com/cockroachdb/cockroach/pkg/server/serverpb._Status_JobStatus_Handler.func1(0x4b67ae0, 0xc05cc554a0, 0x4111980, 0xc0928ac0e0, 0x4111980, 0xc0928ac0e0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/status.pb.go:4470 +0x89 fp=0xc000c79710 sp=0xc000c796c8 pc=0x22499a9
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x4b67ae0, 0xc05cc554a0, 0x4111980, 0xc0928ac0e0, 0xc04197c5e0, 0xc04197c600, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0x9ef fp=0xc000c799b0 sp=0xc000c79710 pc=0x15aebff
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func2(0x4b67ae0, 0xc05cc554a0, 0x4111980, 0xc0928ac0e0, 0xc04197c5e0, 0xc04197c600, 0xc0005f6a78, 0x95fe18, 0x405ffc0, 0xc05cc554a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:223 +0xb6 fp=0xc000c79a18 sp=0xc000c799b0 pc=0x15be386
github.com/cockroachdb/cockroach/pkg/server/serverpb._Status_JobStatus_Handler(0x427e280, 0xc00028d6c0, 0x4b67ae0, 0xc05cc554a0, 0xc0746c01e0, 0xc00032c220, 0x4b67ae0, 0xc05cc554a0, 0xc0928ac0a6, 0xa)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/status.pb.go:4472 +0x14b fp=0xc000c79a88 sp=0xc000c79a18 pc=0x21dca3b
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc0008a2300, 0x4bc17e0, 0xc000796180, 0xc000502300, 0xc000601b60, 0x6dfb380, 0x0, 0x0, 0x0)
```

The job in question was `537299064518082561`. debug.zip too large to be attached, but I grabbed its [system.jobs.txt](https://github.com/cockroachdb/cockroach/files/4326204/system.jobs.txt) and this is from master so it includes the hex repr (this can be a chance to verify that that works for debugging such an issue).

--- a/pkg/jobs/jobs.go
+++ b/pkg/jobs/jobs.go
@@ -566,17 +566,6 @@ func (j *Job) Progress() jobspb.Progress {
 	return j.mu.progress
 }
 
-// ToProto returns the protobuf representation of the job.
-func (j *Job) ToProto() *jobspb.Job {
-	jpb := &jobspb.Job{}
-	j.mu.Lock()
-	defer j.mu.Unlock()
-	jpb.Id = *j.ID()
-	jpb.Payload = protoutil.Clone(&j.mu.payload).(*jobspb.Payload)
-	jpb.Progress = protoutil.Clone(&j.mu.progress).(*jobspb.Progress)
-	return jpb
-}
-
 // Details returns the details from the most recently sent Payload for this Job.
 func (j *Job) Details() jobspb.Details {
 	j.mu.Lock()

--- a/pkg/server/status.go
+++ b/pkg/server/status.go
@@ -35,6 +35,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/build"
 	"github.com/cockroachdb/cockroach/pkg/gossip"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
@@ -2061,14 +2062,23 @@ func (s *statusServer) JobStatus(
 		return nil, err
 	}
 
-	ctx = propagateGatewayMetadata(ctx)
-	ctx = s.AnnotateCtx(ctx)
+	ctx = s.AnnotateCtx(propagateGatewayMetadata(ctx))
 
 	j, err := s.admin.server.jobRegistry.LoadJob(ctx, req.JobId)
 	if err != nil {
 		return nil, err
 	}
-	return &serverpb.JobStatusResponse{
-		Job: j.ToProto(),
-	}, nil
+	res := &jobspb.Job{
+		Payload:  &jobspb.Payload{},
+		Progress: &jobspb.Progress{},
+	}
+	res.Id = *j.ID()
+	// j is not escaping this function and hence is immutable so a shallow copy
+	// is fine. Also we can't really clone the field Payload as it may contain
+	// types that are not supported by Clone().
+	// See https://github.com/cockroachdb/cockroach/issues/46049.
+	*res.Payload = j.Payload()
+	*res.Progress = j.Progress()
+
+	return &serverpb.JobStatusResponse{Job: res}, nil
 }

--- a/pkg/server/status_test.go
+++ b/pkg/server/status_test.go
@@ -29,6 +29,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/build"
 	"github.com/cockroachdb/cockroach/pkg/gossip"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -55,6 +57,7 @@ import (
 	"github.com/gogo/protobuf/proto"
 	"github.com/kr/pretty"
 	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
 )
 
 func getStatusJSONProto(
@@ -1579,4 +1582,64 @@ func TestStatementDiagnosticsCompleted(t *testing.T) {
 		!strings.Contains(json, "statement execution committed the txn") {
 		t.Fatal("statement diagnostics did not capture a trace")
 	}
+}
+
+func TestJobStatusResponse(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ts := startServer(t)
+	defer ts.Stopper().Stop(context.TODO())
+
+	rootConfig := testutils.NewTestBaseContext(security.RootUser)
+	rpcContext := newRPCTestContext(ts, rootConfig)
+
+	url := ts.ServingRPCAddr()
+	nodeID := ts.NodeID()
+	conn, err := rpcContext.GRPCDialNode(url, nodeID, rpc.DefaultClass).Connect(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	client := serverpb.NewStatusClient(conn)
+
+	request := &serverpb.JobStatusRequest{JobId: -1}
+	response, err := client.JobStatus(context.Background(), request)
+	require.Regexp(t, `job with ID -1 does not exist`, err)
+	require.Nil(t, response)
+
+	ctx := context.Background()
+	job, err := ts.JobRegistry().(*jobs.Registry).CreateJobWithTxn(
+		ctx,
+		jobs.Record{
+			Description: "testing",
+			Statement:   "SELECT 1",
+			Username:    "root",
+			Details: jobspb.ImportDetails{
+				Tables: []jobspb.ImportDetails_Table{
+					{
+						Desc: &sqlbase.TableDescriptor{
+							ID: 1,
+						},
+					},
+					{
+						Desc: &sqlbase.TableDescriptor{
+							ID: 2,
+						},
+					},
+				},
+				URIs: []string{"a", "b"},
+			},
+			Progress:      jobspb.ImportProgress{},
+			DescriptorIDs: []sqlbase.ID{1, 2, 3},
+		},
+		nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	request.JobId = *job.ID()
+	response, err = client.JobStatus(context.Background(), request)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.Equal(t, *job.ID(), response.Job.Id)
+	require.Equal(t, job.Payload(), *response.Job.Payload)
+	require.Equal(t, job.Progress(), *response.Job.Progress)
 }

